rust/sbp/src/de.rs

```diff
@@ -174,6 +178,153 @@ impl std::fmt::Display for CrcError {
 
 impl std::error::Error for CrcError {}
 
+// pub struct Framer<R>(FramedRead<R, SbpDecoder>);
+//
+// impl<R> Framer<R> {
+//     pub fn new(reader: R) -> Self {
+//         Self(FramedRead::new(reader, Framer))
+//     }
+// }
+//
+// impl<R: io::Read> Iterator for Framer<R> {
+//     type Item = io::Result<Frame<BytesMut>>;
+//     fn next(&mut self) -> Option<Self::Item> {
+//         self.0.next()
+//     }
+// }
+
+pub struct Framer<R> {
+    inner: FramedRead<R, SbpFramer>,
+}
+
+impl<R: io::Read> Framer<R> {
+    pub fn new(reader: R) -> Self {
+        Self {
+            inner: FramedRead::new(reader, SbpFramer),
+        }
+    }
+}
+
+impl<R: io::Read> Iterator for Framer<R> {
+    type Item = Result<SbpFrame<BytesMut>, Error>;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        self.inner.next()
+    }
+}
+
+// not sure a good naming for this atm
+pub struct SbpDecode<R>(FramedRead<R, SbpFramer>);
+
+impl<R: io::Read> SbpDecode<R> {
+    pub fn new(reader: R) -> Self {
+        SbpDecode(FramedRead::new(reader, SbpFramer))
+    }
+}
+
+impl<R: io::Read> Iterator for SbpDecode<R> {
+    type Item = Result<Sbp, Error>;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        self.0.next().and_then(|frame| {
+            frame.map_or(None, |f| {
+                Some(
+                    f.check_crc()
+                        .map_err(|e| de::Error::from(e))
+                        .and_then(|_| Sbp::from_sbp_frame(f).map_err(|e| de::Error::from(e))),
+                )
+            })
+        })
+    }
+}
+
+#[derive(Debug, Clone, Default)]
+struct SbpFramer;
+
+impl Decoder for SbpFramer {
+    type Item = SbpFrame<BytesMut>;
+    type Error = Error;
+
+    fn decode(&mut self, src: &mut BytesMut) -> Result<Option<Self::Item>, Self::Error> {
+        Ok(SbpFrame::<BytesMut>::parse(src))
+    }
+    fn decode_eof(&mut self, src: &mut BytesMut) -> Result<Option<Self::Item>, Self::Error> {
+        todo!()
+    }
+}
+
+#[derive(Debug)]
+pub struct SbpFrame<B>(B);
```

Comment:
this should just be `Frame`, I'd try to avoid prefixing things with `Sbp`, it's implied by being inside the `sbp` crate. `SbpFramer` has the prefix just to avoid a name collision, but it's private. `FramerImpl` might be a better name

Reply:
in other words I think we should replace the current `Frame` struct in this module with the new one. Usage of the current `Frame` struct is pretty low I think. and anything using it will probably want to utilize the `Framer` so they'll need to be updated anyway